src/com/cdd/bao/template/Schema.java

```diff
@@ -65,15 +65,24 @@ public Group clone(Group parent)
 			for (Group grp : subGroups) dup.subGroups.add(grp.clone(dup));
 			return dup;
 		}
-		public boolean equals(Group other)
+		@Override
+		public boolean equals(Object o)
 		{
+			if (o == null || getClass() != o.getClass()) return false;
+			Group other = (Group)o;
 			if (!name.equals(other.name) || !descr.equals(other.descr) || !groupURI.equals(other.groupURI)) return false;
 			if (assignments.size() != other.assignments.size() || subGroups.size() != other.subGroups.size()) return false;
 			for (int n = 0; n < assignments.size(); n++) if (!assignments.get(n).equals(other.assignments.get(n))) return false;
 			for (int n = 0; n < subGroups.size(); n++) if (!subGroups.get(n).equals(other.subGroups.get(n))) return false;
 			return true;
 		}
 		
+		@Override
+		public int hashCode()
+		{
+			return Objects.hash(name, descr, groupURI, assignments, subGroups);
```

Comment:
`assignments` and `subGroups` are arrays. In this case, you need to make sure that `Assignment` and `Group` also provide adequate methods for `equals` and `hashCode`. Otherwise, java falls back to make comparison based on the memory location. 

Ideally, we should have a test to make sure that the methods work. We can look into this next week.